File: src/main/scala/info/crlog/higgs/protocols/boson/BosonServer.scala
package info.crlog.higgs.protocols.boson

import info.crlog.higgs.RPCServer
import io.netty.channel.ChannelHandlerContext
import java.io.Serializable
import v1.BosonSerializer

/**
 * @author Courtney Robinson <courtney@crlog.info>
 */
class BosonServer(port: Int, host: String = "localhost", compress: Boolean = false)
  extends RPCServer[Message](host, port, compress) {
  val serializer = new BosonSerializer()

  def getArguments(param: Message) = param.arguments.asInstanceOf[Array[AnyRef]]

  def clientCallback(param: Message): String = param.callback

  def decoder() = new BosonDecoder()

  def encoder() = new BosonEncoder()

  def allTopicsKey(): String = ""

  def newResponse(remoteMethodName: String, clientCallbackID: String,
                  response: Option[Serializable], error: Option[Throwable]): Message = {
    val argResponse: Array[Any] = response match {
      case None => Array.empty[Any]
      case Some(arg) => Array(arg)
    }
    val argError: Array[Any] = error match {
      case None => Array.empty[Any]
      case Some(err) => Array(Map("error" -> err.getMessage()))
    }
    val args = argResponse ++ argError
    new Message(clientCallbackID, args)
  }

  def message(context: ChannelHandlerContext, value: Array[Byte]) {
    try {
      val data = serializer.deserialize(value)
      notifySubscribers(context.channel(),
        data.method,
        data
      )
    } catch {
      case e => {
        log.warn("Unable to deserialize message ", e)
        respond(context.channel(),
          new Message("deserialize_error", Array(Map("msg" -> "Incomplete or Invalid Boson message received")))
        )
      }
    }
  }

  override def verifyArgumentType(parameters: Array[AnyRef], methodArguments: Array[Class[_]]): Boolean = {
    if (parameters.length != methodArguments.length) {
      return false //don't invoke
    }
    for (i <- 0 until parameters.length) {
      val methodParam = methodArguments(i)
      val param = parameters(i)
      //if its an array its a bit tricky so types need to be validated carefully
      if (param != null && param.getClass.isArray()) {
        if (!methodParam.isArray()) {
          return false //param received is an array but method isn't expecting one
        }
      }
    }
    true
  }
}

File: higgs-boson/src/main/java/com/fillta/higgs/boson/BosonServer.java
package com.fillta.higgs.boson;

import com.fillta.higgs.RPCServer;
import com.fillta.higgs.boson.serialization.BosonDecoder;
import com.fillta.higgs.boson.serialization.BosonEncoder;
import com.fillta.higgs.boson.serialization.v1.BosonReader;
import com.fillta.higgs.boson.serialization.v1.BosonWriter;
import com.fillta.higgs.events.ChannelMessage;
import com.google.common.base.Optional;
import io.netty.buffer.ByteBuf;
import io.netty.channel.Channel;
import io.netty.channel.ChannelHandlerContext;
import io.netty.channel.ChannelPipeline;

/**
 * @author Courtney Robinson <courtney@crlog.info>
 */
public class BosonServer extends RPCServer<BosonMessage, BosonMessage, ByteBuf> {
	protected boolean compression;
	protected boolean ssl;

	public BosonServer(int port) {
		this(port, false);
		setEnableProtocolSniffing(false);
	}

	public BosonServer(int port, boolean compression) {
		this(port, compression, false);
	}

	public BosonServer(int port, boolean compression, boolean ssl) {
		super(port);
		this.compression = compression;
		this.ssl = ssl;
	}

	public Object[] getArguments(final Class<?>[] argTypes, ChannelMessage<BosonMessage> request) {
		return request.message.arguments;
	}

	protected BosonMessage newResponse(String methodName, ChannelMessage<BosonMessage> request,
	                                   Optional<Object> returns, Optional<Throwable> error) {
		BosonMessage msg = new BosonMessage();
		msg.method = request.message.callback;
		Object[] args = new Object[2];
		args[0] = null;
		if (returns.isPresent()) {
			args[0] = returns.get();
		}
		args[1] = null;
		if (error.isPresent()) {
			args[1] = error.get();
		}
		msg.arguments = args;
		return msg;
	}

	@Override
	public ByteBuf serialize(final Channel ctx, final BosonMessage msg) {
		return new BosonWriter(msg).serialize();
	}

	@Override
	public BosonMessage deserialize(final ChannelHandlerContext ctx, final ByteBuf msg) {
		return new BosonReader(msg).deSerialize();
	}

	@Override
	public String getTopic(final BosonMessage msg) {
		return msg.method;
	}

	@Override
	protected boolean setupPipeline(final ChannelPipeline pipeline) {
		pipeline.addLast("decoder", new BosonDecoder());
		pipeline.addLast("encoder", new BosonEncoder());
		return true;//auto add handler
	}
}
